File: backend/src/main/java/com/eps/procurement/service/DashboardService.java
package com.eps.procurement.service;

import com.eps.procurement.model.*;
import com.eps.procurement.store.DataStore;
import java.util.*;
import org.springframework.stereotype.Service;

/** Aggregates the KPI payloads consumed by each role dashboard. */
@Service
public class DashboardService {

    private final DataStore store;

    public DashboardService(DataStore store) {
        this.store = store;
    }

    private static final Set<String> COMPLETED =
            Set.of("approved", "in_procurement", "delivered", "closed");

    public Map<String, Object> employee(String userId) {
        List<ProcurementRequest> mine = store.requests.stream()
                .filter(r -> r.createdBy.equals(userId))
                .toList();
        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("total", mine.size());
        stats.put("pending", mine.stream().filter(r -> r.status.startsWith("pending")).count());
        stats.put("approved", mine.stream().filter(r -> COMPLETED.contains(r.status)).count());
        stats.put("rejected", mine.stream().filter(r -> r.status.equals("rejected")).count());
        stats.put("inProgress", mine.stream().filter(r -> r.status.equals("in_procurement")).count());

        Map<String, Object> chart = store.chartData();
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("stats", stats);
        body.put("recentRequests", mine.stream().limit(5).toList());
        body.put("monthlyRequests", chart.get("monthlyRequests"));
        body.put("categoryDistribution", chart.get("categoryDistribution"));
        return body;
    }

    public Map<String, Object> manager(String department) {
        List<ProcurementRequest> pending = store.requests.stream()
                .filter(r -> r.status.equals("pending_manager"))
                .filter(r -> department == null || department.isBlank() || r.department.equals(department))
                .toList();
        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("pendingApprovals", pending.size());
        stats.put("approvedThisMonth", countByStatus("approved"));
        stats.put("rejectedThisMonth", countByStatus("rejected"));
        stats.put("totalSpending", store.requests.stream()
                .filter(r -> COMPLETED.contains(r.status))
                .mapToDouble(r -> r.estimatedCost).sum());

        Map<String, Object> chart = store.chartData();
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("stats", stats);
        body.put("pendingRequests", pending);
        body.put("approvalMetrics", chart.get("approvalMetrics"));
        body.put("departmentSpending", chart.get("departmentSpending"));
        return body;
    }

    public Map<String, Object> seniorManager() {
        List<ProcurementRequest> pending = store.requests.stream()
                .filter(r -> r.status.equals("pending_senior_manager"))
                .toList();
        double budget = store.departments.stream().mapToDouble(d -> d.budget).sum();
        double used = store.departments.stream().mapToDouble(d -> d.budgetUsed).sum();

        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("pendingApprovals", pending.size());
        stats.put("approvedThisMonth", countByStatus("approved"));
        stats.put("escalatedRequests", countByStatus("pending_head"));
        stats.put("budgetUtilization", Math.round((used / budget) * 100));

        Map<String, Object> chart = store.chartData();
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("stats", stats);
        body.put("pendingRequests", pending);
        body.put("departmentSpending", chart.get("departmentSpending"));
        body.put("monthlySpending", chart.get("monthlySpending"));
        return body;
    }

    public Map<String, Object> head() {
        double budget = store.departments.stream().mapToDouble(d -> d.budget).sum();
        double used = store.departments.stream().mapToDouble(d -> d.budgetUsed).sum();

        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("totalProcurementSpend", used);
        stats.put("monthlySpend", 2100000);
        stats.put("annualSpend", used);
        stats.put("budgetAllocated", budget);
        stats.put("budgetUsed", used);
        stats.put("budgetRemaining", budget - used);
        stats.put("budgetUtilization", Math.round((used / budget) * 100));
        stats.put("activePOs", store.purchaseOrders.stream().filter(po -> !po.status.equals("closed")).count());
        stats.put("activeSuppliers", store.suppliers.stream().filter(s -> s.status.equals("active")).count());
        stats.put("pendingApprovals", countByStatus("pending_head"));

        Map<String, Object> chart = store.chartData();
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("stats", stats);
        body.put("pendingRequests", store.requests.stream().filter(r -> r.status.equals("pending_head")).toList());
        body.put("categorySpending", chart.get("categoryDistribution"));
        body.put("departmentSpending", chart.get("departmentSpending"));
        body.put("monthlyTrends", chart.get("monthlySpending"));
        body.put("supplierPerformance", chart.get("supplierPerformance"));
        return body;
    }

    public Map<String, Object> procurement() {
        List<ProcurementRequest> approved = store.requests.stream()
                .filter(r -> r.status.equals("approved") || r.status.equals("in_procurement"))
                .toList();
        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("approvedRequests", approved.size());
        stats.put("pendingPOs", store.purchaseOrders.stream()
                .filter(po -> po.status.equals("draft") || po.status.equals("sent")).count());
        stats.put("activePOs", store.purchaseOrders.stream().filter(po -> po.status.equals("accepted")).count());
        stats.put("completedOrders", store.purchaseOrders.stream().filter(po -> po.status.equals("closed")).count());

        Map<String, Object> chart = store.chartData();
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("stats", stats);
        body.put("approvedRequests", approved);
        body.put("supplierPerformance", chart.get("supplierPerformance"));
        body.put("monthlyOrders", chart.get("monthlyRequests"));
        return body;
    }

    public Map<String, Object> finance() {
        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("pendingInvoices", store.payments.stream().filter(p -> p.status.equals("pending")).count());
        stats.put("pendingPayments", store.payments.stream()
                .filter(p -> p.status.equals("processing") || p.status.equals("approved")).count());
        stats.put("paidInvoices", store.payments.stream().filter(p -> p.status.equals("paid")).count());
        stats.put("totalSpending", store.payments.stream()
                .filter(p -> p.status.equals("paid")).mapToDouble(p -> p.amount).sum());

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("stats", stats);
        body.put("payments", store.payments);
        body.put("paymentTrend", store.chartData().get("paymentTrend"));
        return body;
    }

    public Map<String, Object> admin() {
        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("totalUsers", store.users.size());
        stats.put("totalRoles", store.roles.size());
        stats.put("totalDepartments", store.departments.size());
        stats.put("totalCategories", store.categories.size());
        stats.put("totalSuppliers", store.suppliers.size());
        stats.put("totalAuditLogs", store.auditLogs.size());

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("stats", stats);
        body.put("recentAuditLogs", store.auditLogs.stream().limit(5).toList());
        return body;
    }

    private long countByStatus(String status) {
        return store.requests.stream().filter(r -> r.status.equals(status)).count();
    }
}

File: front-end/src/services/dashboardService.js
import { apiCall, isMockMode } from './apiConfig';
import { requests, suppliers, purchaseOrders, payments, users, departments, chartData, notifications as mockNotifications, auditLogs, softwareLicenses, goodsReceiptNotes, categories, roles, quotations, approvalRules } from '../data/mockData';

// ============================================
// Dashboard Service
// ============================================

/**
 * GET /dashboard/employee?userId=string
 * 
 * Response (200):
 * {
 *   "stats": { "total": n, "pending": n, "approved": n, "rejected": n, "inProgress": n },
 *   "recentRequests": [...],
 *   "monthlyRequests": [{ "month": "string", "requests": n }],
 *   "categoryDistribution": [{ "name": "string", "value": n }]
 * }
 */
export const getEmployeeDashboard = async (userId) => {
  // return apiCall(`/dashboard/employee?userId=${userId}`);
  if (isMockMode()) {
    const userRequests = requests.filter(r => r.createdBy === userId);
    return {
      stats: {
        total: userRequests.length,
        pending: userRequests.filter(r => r.status.startsWith('pending')).length,
        approved: userRequests.filter(r => r.status === 'approved' || r.status === 'in_procurement' || r.status === 'delivered' || r.status === 'closed').length,
        rejected: userRequests.filter(r => r.status === 'rejected').length,
        inProgress: userRequests.filter(r => r.status === 'in_procurement').length,
      },
      recentRequests: userRequests.slice(0, 5),
      monthlyRequests: chartData.monthlyRequests,
      categoryDistribution: chartData.categoryDistribution,
    };
  }
};

/**
 * GET /dashboard/manager?department=string
 * 
 * Response (200):
 * {
 *   "stats": { "pendingApprovals": n, "approvedThisMonth": n, "rejectedThisMonth": n, "totalSpending": n },
 *   "pendingRequests": [...],
 *   "approvalMetrics": [{ "month": "string", "avgDays": n, "approvalRate": n }],
 *   "departmentSpending": [...]
 * }
 */
export const getManagerDashboard = async (department) => {
  // return apiCall(`/dashboard/manager?department=${department}`);
  if (isMockMode()) {
    const pending = requests.filter(r => r.status === 'pending_manager');
    return {
      stats: {
        pendingApprovals: pending.length,
        approvedThisMonth: 8,
        rejectedThisMonth: 2,
        totalSpending: 1250000,
      },
      pendingRequests: pending,
      approvalMetrics: chartData.approvalMetrics,
      departmentSpending: chartData.departmentSpending,
    };
  }
};

/**
 * GET /dashboard/senior-manager
 */
export const getSeniorManagerDashboard = async () => {
  // return apiCall('/dashboard/senior-manager');
  if (isMockMode()) {
    const pending = requests.filter(r => r.status === 'pending_senior_manager');
    return {
      stats: {
        pendingApprovals: pending.length,
        approvedThisMonth: 12,
        escalatedRequests: 3,
        budgetUtilization: 68,
      },
      pendingRequests: pending,
      departmentSpending: chartData.departmentSpending,
      monthlySpending: chartData.monthlySpending,
    };
  }
};

/**
 * GET /dashboard/head
 * 
 * Response (200):
 * {
 *   "stats": {
 *     "totalProcurementSpend": n, "monthlySpend": n, "annualSpend": n,
 *     "budgetAllocated": n, "budgetUsed": n, "budgetRemaining": n,
 *     "budgetUtilization": n, "activePOs": n, "activeSuppliers": n
 *   },
 *   "categorySpending": [...], "departmentSpending": [...],
 *   "monthlyTrends": [...], "supplierPerformance": [...]
 * }
 */
export const getHeadDashboard = async () => {
  // return apiCall('/dashboard/head');
  if (isMockMode()) {
    const totalBudget = departments.reduce((s, d) => s + d.budget, 0);
    const totalUsed = departments.reduce((s, d) => s + d.budgetUsed, 0);
    return {
      stats: {
        totalProcurementSpend: totalUsed,
        monthlySpend: 2100000,
        annualSpend: totalUsed,
        budgetAllocated: totalBudget,
        budgetUsed: totalUsed,
        budgetRemaining: totalBudget - totalUsed,
        budgetUtilization: Math.round((totalUsed / totalBudget) * 100),
        activePOs: purchaseOrders.filter(po => po.status !== 'closed').length,
        activeSuppliers: suppliers.filter(s => s.status === 'active').length,
      },
      categorySpending: chartData.categoryDistribution,
      departmentSpending: chartData.departmentSpending,
      monthlyTrends: chartData.monthlySpending,
      supplierPerformance: chartData.supplierPerformance,
    };
  }
};

/**
 * GET /dashboard/procurement
 */
export const getProcurementDashboard = async () => {
  // return apiCall('/dashboard/procurement');
  if (isMockMode()) {
    return {
      stats: {
        approvedRequests: requests.filter(r => r.status === 'approved' || r.status === 'in_procurement').length,
        pendingPOs: purchaseOrders.filter(po => po.status === 'draft' || po.status === 'sent').length,
        activePOs: purchaseOrders.filter(po => po.status === 'accepted').length,
        completedOrders: purchaseOrders.filter(po => po.status === 'closed').length,
      },
      approvedRequests: requests.filter(r => r.status === 'approved' || r.status === 'in_procurement'),
      supplierPerformance: chartData.supplierPerformance,
      monthlyOrders: chartData.monthlyRequests,
    };
  }
};

/**
 * GET /dashboard/finance
 */
export const getFinanceDashboard = async () => {
  // return apiCall('/dashboard/finance');
  if (isMockMode()) {
    return {
      stats: {
        pendingInvoices: payments.filter(p => p.status === 'pending').length,
        pendingPayments: payments.filter(p => p.status === 'processing' || p.status === 'approved').length,
        paidInvoices: payments.filter(p => p.status === 'paid').length,
        totalSpending: payments.filter(p => p.status === 'paid').reduce((s, p) => s + p.amount, 0),
      },
      payments: payments,
      paymentTrend: chartData.paymentTrend,
    };
  }
};

/**
 * GET /dashboard/admin
 */
export const getAdminDashboard = async () => {
  // return apiCall('/dashboard/admin');
  if (isMockMode()) {
    return {
      stats: {
        totalUsers: users.length,
        totalRoles: roles.length,
        totalDepartments: departments.length,
        totalCategories: categories.length,
        totalSuppliers: suppliers.length,
        totalAuditLogs: auditLogs.length,
      },
    };
  }
};
